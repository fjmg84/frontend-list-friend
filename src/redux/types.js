export const TYPES = {
  LOAD: "load/friends",
  ERROR_ACTIVE: "error/active",
};
