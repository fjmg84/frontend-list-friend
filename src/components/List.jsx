import React from "react";
import PropTypes from "prop-types";
import { useNavigate } from "react-router-dom";
import image from "../photo.jpg";

export const List = ({ friends }) => {
  const navigate = useNavigate();

  return (
    <ul>
      {friends.map((friend) => {
        return (
          <li key={friend.id}>
            <div className="card">
              <div className="card-img">
                <img src={image} alt={friend.img} className="small" />
                <span
                  className={`availability eclipse ${
                    friend.available ? "available" : "not-available"
                  }`}
                ></span>
              </div>

              <div className="card-header ">
                <div className="card-header-name">
                  <span>{`${friend.first_name} ${friend.last_name}`}</span>
                </div>
                <div className="card-header-status">
                  <span>{friend.status}</span>
                </div>
              </div>

              <div className="card-actions">
                <button
                  className="details-button"
                  onClick={() => navigate(`./show/${friend.id}`)}
                  disabled={!friend.available}
                >
                  Details
                </button>
              </div>
            </div>
          </li>
        );
      })}
    </ul>
  );
};

List.protoTypes = {
  friends: PropTypes.array,
};

List.propTypes = {
  friends: PropTypes.array.isRequired,
};
