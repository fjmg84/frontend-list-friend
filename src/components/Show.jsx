import React, { useEffect, useState } from "react";
import { useParams, useNavigate } from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";
import { findOneFriend } from "../redux/actions/friends";
import { setAlert } from "../redux/actions/alerts";
import img from "../photo.jpg";

export const Show = () => {
  let { id } = useParams();
  const dispatch = useDispatch();
  const { status, msg } = useSelector((state) => state.alertsReducer);
  const friends = useSelector((state) => state.friendsReducer);
  const [statusOfTheFriend, setStatusOfTheFriend] = useState(() => {
    return friends?.filter((f) => f.id === +id)[0].status;
  });

  const [view, setView] = useState(true);
  const [friend, setFriend] = useState({});
  const bio = "";
  const navigate = useNavigate();

  useEffect(() => {
    dispatch(setAlert(true, "loading..."));
    findOneFriend(id)
      .then((friend) => {
        setFriend(friend);
        dispatch(setAlert(false, undefined));
      })
      .catch((err) => {
        console.error(err);
        dispatch(setAlert(true, "Up!!! Sorry. Not Found data of this friend."));
      });
  }, [id, dispatch]);

  const handleView = () => {
    setView(!view);
  };

  const info = () => {
    return (
      <>
        <div className="article col">
          <p className="title">Bio:</p>
          <div className="text">{friend.bio}</div>
        </div>
        <div className="divider"></div>

        <div className="article row">
          <p className="title">Phone:</p>
          <div className="">{friend.phone}</div>
        </div>
        <div className="divider"></div>

        <div>
          <div className="article row">
            <p className="title">Address</p>
            <div>{friend.address_1}</div>
          </div>

          <div className="article row">
            <p className="title">City</p>
            <div>{friend.city}</div>
          </div>

          <div className="article row">
            <p className="title">Zipcode</p>
            <div>{friend.zipcode}</div>
          </div>

          <div className="article row">
            <p className="title">State</p>
            <div>{friend.state}</div>
          </div>
        </div>
      </>
    );
  };

  const photos = () => {
    return (
      <div className="article">
        <div className="gallery">
          <img src={img} />
          <img src={img} />
          <img src={img} />

          <img src={img} />
          <img src={img} />
          <img src={img} />

          <img src={img} />
          <img src={img} />
          <img src={img} />
        </div>
      </div>
    );
  };

  const handleNavegation = () => {
    navigate(-1);
  };

  return (
    <div className="box">
      <div className="arrow">
        <span className="fill vector" onClick={() => handleNavegation()}></span>
      </div>

      <div className="body">
        <div className="header">
          <h1>Friend</h1>
          {status && <div className="error">{msg}</div>}
        </div>

        <div className="main">
          <div className="show-card">
            <div className="show-card-img">
              <img src={img} alt={friend.img} className="medium" />
              <span className="availability eclipse available"></span>
            </div>
            <div className="show-card-content">
              <p>{`${friend.first_name} ${friend.last_name}`}</p>
              <div className="show-card-header-status">
                <span>{statusOfTheFriend}</span>
              </div>

              <div className="show-card-box-data">
                <div className="nav">
                  <div className={`tab1 ${view ? "active" : "inactive"}`}>
                    <span onClick={() => handleView()}>Info</span>
                    <span className={`${view ? "line" : ""}`}></span>
                  </div>
                  <div className={`tab2 ${!view ? "active" : "inactive"}`}>
                    <span onClick={() => handleView()}>Photos</span>
                    <span className={`${!view ? "line" : ""}`}></span>
                  </div>
                </div>

                <div className="section">{view ? info() : photos()}</div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};
