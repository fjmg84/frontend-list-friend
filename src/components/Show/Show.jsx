import React, { useEffect, useState } from 'react'
import { useParams, useNavigate } from 'react-router-dom'
import { useDispatch, useSelector } from 'react-redux'
import { MdKeyboardBackspace } from 'react-icons/md'
import { findOneFriend } from '../../redux/actions/friends'
import { setAlert } from '../../redux/actions/alerts'
import { Info } from './Details/Info'
import { Photos } from './Details/Photos'
import { avatar } from '../../images/index'
import { Spinner } from '../Helps/Spinner'

export const Show = () => {
	let { id } = useParams()

	const dispatch = useDispatch()
	const { status, msg, classAlert } = useSelector(
		(state) => state.alertsReducer,
	)
	const friends = useSelector((state) => state.friendsReducer)
	const [view, setView] = useState(true)
	const [friend, setFriend] = useState({})
	const statusOfFriend = friends.filter((friend) => friend.id === +id)[0]

	const navigate = useNavigate()

	useEffect(() => {
		dispatch(setAlert(true, 'Loading...', 'loading'))
		findOneFriend(id)
			.then((friend) => {
				setFriend(friend)
				dispatch(setAlert(false, undefined))
			})
			.catch((err) => {
				console.error(err)
				dispatch(
					setAlert(
						true,
						'Up!!! Sorry. Not Found Data For This Friend :(.',
						'error',
					),
				)
			})
	}, [id, dispatch])

	return (
		<div className="box">
			<div className="icons icons-blue arrow">
				<span className="fill vector" onClick={() => navigate('/')}>
					<MdKeyboardBackspace size={24} />
				</span>
			</div>

			<div className="body">
				<div className="header">
					<h1>Friend</h1>
					{status && <div className={classAlert}>{msg}</div>}
				</div>

				<div className="main">
					{friend?.first_name ? (
						<div className="show-card">
							<div className="show-card-img">
								<img src={avatar} alt={friend.img} className="medium" />
								<span className="availability eclipse available"></span>
							</div>

							<div className="show-card-content">
								<p className="details-name">{`${friend.first_name} ${friend?.last_name}`}</p>
								<div className="show-card-header-status">
									<span className="details-status">
										{statusOfFriend?.status}
									</span>
								</div>

								<div className="show-card-box-data">
									<div className="nav">
										<div className={`tab ${view ? 'active' : 'inactive'}`}>
											<span onClick={() => setView(!view)}>Info</span>
											<span className={`${view ? 'line' : ''}`}></span>
										</div>
										<div className={`tab ${!view ? 'active' : 'inactive'}`}>
											<span onClick={() => setView(!view)}>Photos</span>
											<span className={`${!view ? 'line' : ''}`}></span>
										</div>
									</div>

									<div className="section">
										{view ? (
											<Info friend={friend} />
										) : (
											<Photos photos={friend.photos} />
										)}
									</div>
								</div>
							</div>
						</div>
					) : (
						<Spinner />
					)}
				</div>
			</div>
		</div>
	)
}
