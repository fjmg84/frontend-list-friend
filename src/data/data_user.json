{
    "id": 6,
    "img": "https://s3.amazonaws.com/uifaces/faces/twitter/walterstephanie/128.jpg",
    "first_name": "Steph",
    "last_name": "Walters",
    "phone": "(820) 289-1818",
    "address_1": "5190 Center Court Drive",
    "city": "Spring",
    "state": "TX",
    "zipcode": "77370",
    "bio": "I'm very choosy. I'm also very suspicious, very irrational and I have a very short temper. I'm also extremely jealous and slow to forgive. Just so you know.",
    "photos": [
        "/static/media/photo_1.65b44377ca3ba76ac460.jpg",
        "/static/media/photo_2.f94104513f3ac56066cc.jpg",
        "/static/media/photo_3.a617d38fcaf25508172f.jpg"
    ],
    "statuses": [
        "Developing something amazing",
        "This could be interesting....",
        "Man, life is so good",
        "There is nothing quite like a good friend",
        "Take a look around you, everything is awesome",
        "What is the point of all of this"
    ],
    "available": true
}