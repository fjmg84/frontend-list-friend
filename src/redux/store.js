import { createStore, combineReducers } from "redux";
import { friendsReducer } from "../reducer/friends";
import { errorsReducer } from "../reducer/errors";

const rootReducers = combineReducers({
  friendsReducer,
  errorsReducer,
});

export const store = createStore(
  rootReducers,
  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
);
