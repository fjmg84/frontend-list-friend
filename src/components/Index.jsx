import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { getFriends } from "../redux/actions/friends";
import { setAlert } from "../redux/actions/alerts";
import { TYPES } from "../redux/types";
import { List } from "./List";

export const Index = () => {
  const dispatch = useDispatch();
  const { status, msg, classAlert } = useSelector(
    (state) => state.alertsReducer
  );

  useEffect(() => {
    dispatch(setAlert(true, "Loading friends...", "loading"));
    getFriends()
      .then((friends) => {
        dispatch({ type: TYPES.LOAD, payload: friends });
        dispatch(setAlert(false, undefined));
      })
      .catch((err) => {
        console.log(err);
        dispatch(setAlert(true, "Up!! Sorry. Not loading Friends :(", "error"));
      });
  }, [dispatch]);

  return (
    <div className="box">
      <div className="body">
        <div className="header">
          <h1>Friends</h1>
          {status && <div className={classAlert}>{msg}</div>}
        </div>

        <div className="main">
          <List />
        </div>
      </div>
    </div>
  );
};
