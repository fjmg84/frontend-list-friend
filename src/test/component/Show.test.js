import React from "react";
import { act } from "react-dom/test-utils";
import { MemoryRouter } from "react-router-dom";
import { Provider } from "react-redux";
import { mount, render, shallow } from "enzyme";

import { Show } from "../../components/Show";
import { store } from "../../redux/store";
import { TYPES } from "../../redux/types";
import data_user from "../../data_user.json";
import data from "../../data.json";

describe("testing component <Show/>", () => {
  test("snapshot to component <Show/>", () => {
    const wrapper = shallow(
      <MemoryRouter>
        <Provider store={store}>
          <Show />
        </Provider>
      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
  });

  test("render tag alert in component", async () => {
    store.dispatch({
      type: TYPES.CREATE_ALERT,
      payload: {
        status: true,
        msg: "Testing to component",
        classAlert: "loading",
      },
    });
    await act(async () => {
      const wrapper = mount(
        <MemoryRouter>
          <Provider store={store}>
            <Show />
          </Provider>
        </MemoryRouter>
      );

      expect(wrapper.find(".loading").text()).toEqual("Testing to component");
    });
  });

  /* test("show data in component", async () => {
    const wrapper = render(
      <MemoryRouter>
        <Provider store={store}>
          <Show />
        </Provider>
      </MemoryRouter>
    );

    console.log(store.getState());
  }); */
});
