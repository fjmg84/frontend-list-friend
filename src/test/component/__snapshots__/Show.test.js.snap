// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing component <Show/> snapshot to component <Show/> 1`] = `
<Router
  location={
    Object {
      "hash": "",
      "key": "5x9zsih8",
      "pathname": "/",
      "search": "",
      "state": null,
    }
  }
  navigationType="POP"
  navigator={
    Object {
      "action": "POP",
      "back": [Function],
      "block": [Function],
      "createHref": [Function],
      "forward": [Function],
      "go": [Function],
      "index": 0,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "key": "5x9zsih8",
        "pathname": "/",
        "search": "",
        "state": null,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <Provider
    store={
      Object {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <Show />
  </Provider>
</Router>
`;
